c++/src/capnp/serialize-async.c++

```diff
@@ -541,4 +545,181 @@ kj::Promise<MessageReaderAndFds> MessageStream::readMessage(
+// "retaining" a message
+
+class BufferedAsyncMessageStream final : public BufferedMessageStream {
+public:
+  explicit BufferedAsyncMessageStream(kj::AsyncIoStream& stream,
+      BufferedStreamOptions options)
+      : buffer(options.bufferSizeInWords * sizeof(word), options.maxFdsPerMessage),
+        stream(&stream), options(options) {}
+  explicit BufferedAsyncMessageStream(kj::AsyncCapabilityStream& stream,
+      BufferedStreamOptions options)
+      : buffer(options.bufferSizeInWords * sizeof(word), options.maxFdsPerMessage),
+        stream(&stream), options(options) {}
+
+  // Implements MessageStream
+  kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessage(
+      kj::ArrayPtr<kj::AutoCloseFd> fdSpace,
+      ReaderOptions options = ReaderOptions(), kj::ArrayPtr<word> scratchSpace = nullptr) override;
+  kj::Promise<void> writeMessage(
+      kj::ArrayPtr<const int> fds,
+      kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) override;
+  kj::Promise<void> writeMessages(
+      kj::ArrayPtr<kj::ArrayPtr<const kj::ArrayPtr<const word>>> messages) override;
+  kj::Maybe<int> getSendBufferSize() override;
+  kj::Promise<void> end() override;
+
+  // Implements BufferedMessageStream
+  void retainLastMessageLongTerm() override;
+
+private:
+  kj::FdRangeInputBuffer buffer;
+  kj::Maybe<kj::Own<kj::RetainableFdBufferRange>> lastRead;
+
+  kj::OneOf<kj::AsyncIoStream*, kj::AsyncCapabilityStream*> stream;
+  BufferedStreamOptions options;
+};
+
+kj::Own<BufferedMessageStream> makeBufferedStream(kj::AsyncIoStream& ioStream,
+    BufferedStreamOptions options) {
+  return kj::heap<BufferedAsyncMessageStream>(ioStream, options);
+}
+
+kj::Own<BufferedMessageStream> makeBufferedStream(kj::AsyncCapabilityStream& capStream,
+    BufferedStreamOptions options) {
+  return kj::heap<BufferedAsyncMessageStream>(capStream, options);
+}
+
+kj::ArrayPtr<const word> asWords(kj::ArrayPtr<const byte> bytes) {
+  return kj::ArrayPtr<const word>(
+      reinterpret_cast<const word*>(bytes.begin()), bytes.size() / sizeof(word));
+}
+
+namespace {
+  kj::Promise<kj::Own<kj::RetainableFdBufferRange>> finishMessage(
+      kj::OneOf<kj::AsyncIoStream*, kj::AsyncCapabilityStream*> stream,
+      kj::Own<kj::SplittableFdBufferRange> range) {
+    auto bytes = range->getBytes();
+    auto words = asWords(bytes);
+
+    auto expectedSize = expectedSizeInWordsFromPrefix(words);
+    auto expectedSizeInBytes = expectedSize * sizeof(word);
+
+    if (expectedSize <= words.size()) {
+      // Complete message, split it off
+      return range->split(expectedSizeInBytes);
+    } else {
+      // Incomplete message, finish it
+      return range->appendFromStream(stream, expectedSizeInBytes - bytes.size())
+          .then([stream, range = kj::mv(range)]() mutable {
+        return finishMessage(stream, kj::mv(range));
+      });
+    }
+  }
+}
+
+kj::Promise<kj::Maybe<MessageReaderAndFds>> BufferedAsyncMessageStream::tryReadMessage(
```

Comment:
This is pretty much the only function that doesn't delegate to the underlying stream. I think we should be able to do something like `return buffer.tryReadMessage(...);`. I'd guess you don't do this because the `FdRangeInputBuffer` is in kj and that would require knowing about `capnp::MessageReaderAndFds`. I really think it's worth moving `buffering.c++` to capnp if so.

I think it's worth a lot to be able to view the entire read message implementation in one place instead of splitting it up like this.